File: algorithm/src/com/ssafy/hong/boj10745.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <string>
#include <queue>

using namespace std;
const int MAX=27, M=100002;
char str[M];


struct Trie{
	Trie* next[MAX], *fail;
	bool finish;
	int length;
	Trie(){fill(next,next+MAX,nullptr); fail=nullptr; finish=false; length=0;}
	~Trie(){for(int i=0; i<MAX; i++)if(next[i])delete next[i];}

	void insert(string str){
		Trie* temp = this;
		for(auto c : str){
			int idx = c-'a';
			if(temp->next[idx]==nullptr)temp->next[idx]=new Trie();
			temp=temp->next[idx];
		}
		temp->finish=true;
		temp->length=str.size();
	}
};

void makeFail(Trie* root){
	queue<Trie*> q;
	q.push(root);
	root->fail=root;

	while(!q.empty()){
		Trie * curr=q.front(); q.pop();
		for(int i=0; i<MAX; i++){
			Trie * next = curr->next[i];
			Trie * fail = curr->fail;
			if(next==nullptr)continue;
			if(curr==root)next->fail=root;
			else{
				while(fail->next[i] == nullptr && fail != root) fail=fail->fail;
				if(fail->next[i]) fail=fail->next[i];
				next->fail=fail;
			}
			if(next->fail->finish){
				next->finish=true;
				next->length=next->fail->length;
			}
			q.push(next);
		}
	}
	for(int i=0; i<MAX; i++){
		if(root->next[i])q.push(root->next[i]);
		else root->next[i]=root;
	}
	while(!q.empty()){
		Trie* curr = q.front();
		q.pop();
		for(int i=0; i<MAX; i++){
			if(curr->next[i])q.push(curr->next[i]);
			else curr->next[i]=curr->fail->next[i];
		}
	}
}

void getAns(Trie* root, string& s){
	int idx=0, size =s.size();
	vector<Trie*> vTrie(size);
	
	Trie* curr = root;
	vTrie[idx++] = root;

	for(int i=0; i<size; i++){
		int indx = s[i]-'a';
		while(curr->next[indx] == nullptr && curr!=root){
			curr=curr->fail;
		}
		if(curr->next[indx]){
			curr=curr->next[indx];
		}

		vTrie[idx] = curr;
		str[idx++-1] = s[i];

		if(curr->finish){
			idx-=curr->length;
			curr=vTrie[idx-1];
		}
	}
	str[idx-1]='\0';
	cout<<str;
}
int main(void){
	cin.tie(0);
	ios_base::sync_with_stdio(0);

	Trie* root = new Trie();
	int N;
	string S,t; 
	cin>>S>>N;

	for(int i=0; i<N; i++){
		cin>>t;
		root->insert(t);
	}
	makeFail(root);
	getAns(root,S);
	return 0;
}

File: algorithm/src/com/ssafy/hong/Boj10745.java
package com.ssafy.hong;

import java.io.IOException;
import java.util.*;

public class Boj10745 {
    /**
     * [BOJ: 10745] Censoring
     * 길이 최대 10만인 S 잡지의 text에서
     * N개의 단어들을 모두 제거하라. 단, 제거 후 생기는 지워야 할 단어들도 제거해야 한다.
     * 
     * input:   S     잡지 text
     *          N     지워야 할 단어 수
     *          str1  단어
     *          ..
     * output:  D   모든 단어 제거 후의 문자열
     * @param args
     * @throws IOException
     */
    public static void main(String[] args)throws IOException {
        Scanner sc = new Scanner(System.in);
        String S = sc.next();
        Trie root = new Trie();
        int N = sc.nextInt();
        for(int i=0; i<N; i++){
            String D = sc.next();
            root.insert(D);
        }

        setFailure(root);
        Trie curr = root;
        Trie[] tries = new Trie[100002];
        char[] ans = new char[100002];
        int idx=0;
        tries[idx++]=root;
        for(int i=0; i<S.length(); i++){
            char t = S.charAt(i);
            curr = curr.next[t-'a'];

            ans[idx-1]=t;
            tries[idx++]=curr;

            if(curr.finish){
                idx-=curr.length;
                curr=tries[idx-1];
            }
        }
        StringBuilder sb = new StringBuilder();
        for(int i=0; i<idx-1; i++) sb.append(ans[i]);
        System.out.println(sb.toString());
    }

    static void setFailure(Trie root){
        Queue<Trie> q = new LinkedList<>();
        q.add(root);
        root.fail=root;

        while(q.size()>0){
            Trie curr = q.poll();

            for(int i=0; i<26; i++){
                Trie next = curr.next[i];
                Trie fail = curr.fail;
                if(next==null)continue;
                if(curr==root)next.fail=root;
                else{
                    while(fail != root && fail.next[i] == null){
                        fail=fail.fail;
                    }
                    if(fail.next[i]!=null)fail=fail.next[i];
                    next.fail=fail;
                }
                if(fail.finish){
                    next.finish=true;
                    next.length=fail.length;
                }
                q.add(next);
            }
        }

        for(int i=0; i<26; i++){
            if(root.next[i] != null) q.add(root.next[i]);
            else root.next[i]=root;
        }
        while(q.size()>0){
            Trie curr = q.poll();

            for(int i=0; i<26; i++){
                if(curr.next[i] != null)q.add(curr.next[i]);
                else curr.next[i]=curr.fail.next[i];
            }
        }
    }


    static class Trie{
        Trie[] next = new Trie[26];
        Trie fail;
        boolean finish;
        int length;

        void insert(String str){
            Trie temp = this;
            for(int i=0; i<str.length(); i++){
                int idx = str.charAt(i)-'a';
                if(temp.next[idx]==null) temp.next[idx]=new Trie();
                temp = temp.next[idx];
            }
            temp.finish=true;
            temp.length=str.length();
        }
    }

}
